llvm/lib/SYCLLowerIR/LowerWGLocalMemory.cpp

```diff
@@ -110,15 +152,171 @@ static bool allocaWGLocalMemory(Module &M) {
   }
 
   // Remove __sycl_allocateLocalMemory declaration.
-  assert(ALMFunc->use_empty() && "__sycl_allocateLocalMemory is still in use");
-  ALMFunc->eraseFromParent();
+  assert(LocalMemAllocFunc->use_empty() &&
+         "local mem allocation function is still in use");
+  LocalMemAllocFunc->eraseFromParent();
+}
+
+static bool allocaWGLocalMemory(Module &M) {
+  Function *ALMFunc = M.getFunction(SYCL_ALLOCLOCALMEM_CALL);
+  if (!ALMFunc)
+    return false;
+
+  assert(ALMFunc->isDeclaration() && "should have declaration only");
+
+  lowerLocalMemCall(ALMFunc,
+                    [&](CallInst *CI) { lowerAllocaLocalMemCall(CI, M); });
+
+  return true;
+}
+
+// For dynamic memory we have 2 case:
+//   - Direct for CUDA/HIP: we create a placeholder and set the memory on launch
+//   - Indirect for OpenCL/Level0: we create a shared value holding the pointer
+//   to the buffer passed as argument
+static bool dynamicWGLocalMemory(Module &M) {
+  Function *DLMFunc = M.getFunction(SYCL_DYNAMIC_LOCALMEM_CALL);
+  if (!DLMFunc)
+    return false;
+
+  GlobalVariable *LocalMemArrayGV =
+      M.getGlobalVariable(DYNAMIC_LOCALMEM_GV, true);
+  Triple TT(M.getTargetTriple());
+  unsigned LocalAS = DLMFunc->getReturnType()->getPointerAddressSpace();
+
+  if (!LocalMemArrayGV) {
+    assert(DLMFunc->isDeclaration() && "should have declaration only");
+    Type *LocalMemArrayTy =
+        TT.isSPIROrSPIRV()
+            ? static_cast<Type *>(PointerType::get(M.getContext(), LocalAS))
+            : static_cast<Type *>(
+                  ArrayType::get(Type::getInt8Ty(M.getContext()), 0));
+    LocalMemArrayGV = new GlobalVariable(
+        M,               // module
+        LocalMemArrayTy, // type
+        false,           // isConstant
+        TT.isSPIROrSPIRV() ? GlobalValue::LinkOnceODRLinkage
+                           : GlobalValue::ExternalLinkage, // Linkage
+        TT.isSPIROrSPIRV() ? UndefValue::get(LocalMemArrayTy)
+                           : nullptr,   // Initializer
+        DYNAMIC_LOCALMEM_GV,            // Name prefix
+        nullptr,                        // InsertBefore
+        GlobalVariable::NotThreadLocal, // ThreadLocalMode
+        LocalAS                         // AddressSpace
+    );
+    if (!TT.isSPIROrSPIRV())
+      LocalMemArrayGV->setAlignment(Align{128});
```

Comment:
Why 128? Feels like a magic number.

Reply:
That's the max alignment for NV, and that's also required the TMA descriptors. Adding this in source is a bit awkward ATM, so I left it as an hardcoded value until we revisit this.

I change it to a `constexpr` variable to be more explicit of the intent here.